avocado/core/nrunner.py

```diff
@@ -0,0 +1,225 @@
+import argparse
+import io
+import json
+import multiprocessing
+import subprocess
+import time
+import unittest
+
+#: The amount of time between each internal status check
+RUNNER_RUN_CHECK_INTERVAL = 0.01
+
+#: The amount of time between a status report from a runner that
+#: performs its work asynchronously
+RUNNER_RUN_STATUS_INTERVAL = 0.5
+
+
+class Runnable:
+    """
+    Describes an entity that be executed in the context of a task
+
+    A instance of :class:`BaseRunner` is the entity that will actually
+    execute a runnable.
+    """
+    def __init__(self, kind, uri=None, *args, **kwargs):
+        self.kind = kind
+        self.uri = uri
+        self.args = args
+        self.kwargs = kwargs
+
+
+def runnable_from_recipe(recipe_path):
+    """
+    Returns a runnable from a runnable recipe file
+    """
+    with open(recipe_path) as recipe_file:
+        recipe = json.load(recipe_file)
+    return Runnable(recipe.get('kind'),
+                    recipe.get('uri'),
+                    *recipe.get('args', ()))
+
+
+class BaseRunner:
+    """
+    Base interface for a Runner
+    """
+
+    def __init__(self, runnable):
+        self.runnable = runnable
+
+    def run(self):
```

Comment:
Do you think in provide exception classes for runners?

Reply:
That may be end up being necessary... but the main aspect is that, given that a runner will (usually, and on this version always) run a different process, the main communication channel for exceptional situations would be the yield(ed)() data.

Reply:
ok. Yes, since it is a experimental feature we can add exception classes (if ever needed) later.